src/app/core/end-user-agreement/end-user-agreement.service.ts

```diff
@@ -0,0 +1,109 @@
+import { hasValue } from '../../shared/empty.util';
+import { cloneDeep } from 'lodash';
+import { Metadata } from '../shared/metadata.utils';
+import { EPersonDataService } from '../eperson/eperson-data.service';
+import { getSucceededRemoteData } from '../shared/operators';
+
+export const END_USER_AGREEMENT_COOKIE = 'hasAgreedEndUser';
+export const END_USER_AGREEMENT_METADATA_FIELD = 'dspace.agreements.end-user';
+
+/**
+ * Service for checking and managing the status of the current end user agreement
+ */
+@Injectable()
+export class EndUserAgreementService {
+
+  constructor(protected cookie: CookieService,
+              protected authService: AuthService,
+              protected ePersonService: EPersonDataService) {
+  }
+
+  /**
+   * Whether or not either the cookie was accepted or the current user has accepted the End User Agreement
+   * @param acceptedWhenAnonymous Whether or not the user agreement should be considered accepted if the user is
+   *                              currently not authenticated (anonymous)
+   */
+  hasCurrentUserOrCookieAcceptedAgreement(acceptedWhenAnonymous: boolean): Observable<boolean> {
+    if (this.isCookieAccepted()) {
+      return observableOf(true);
+    } else {
+      return this.hasCurrentUserAcceptedAgreement(acceptedWhenAnonymous);
+    }
+  }
+
+  /**
+   * Whether or not the current user has accepted the End User Agreement
+   * @param acceptedWhenAnonymous Whether or not the user agreement should be considered accepted if the user is
+   *                              currently not authenticated (anonymous)
+   */
+  hasCurrentUserAcceptedAgreement(acceptedWhenAnonymous: boolean): Observable<boolean> {
+    return this.authService.isAuthenticated().pipe(
+      switchMap((authenticated) => {
+        if (authenticated) {
+          return this.authService.getAuthenticatedUserFromStore().pipe(
+            map((user) => hasValue(user) && user.hasMetadata(END_USER_AGREEMENT_METADATA_FIELD) && user.firstMetadata(END_USER_AGREEMENT_METADATA_FIELD).value === 'true')
+          );
+        } else {
+          return observableOf(acceptedWhenAnonymous);
+        }
+      })
+    );
+  }
+
+  /**
+   * Set the current user's accepted agreement status
+   * When a user is authenticated, set his/her metadata to the provided value
+   * When no user is authenticated, set the cookie to the provided value
+   * @param accepted
+   */
+  setUserAcceptedAgreement(accepted: boolean): Observable<boolean> {
+    return this.authService.isAuthenticated().pipe(
+      switchMap((authenticated) => {
+        if (authenticated) {
+          return this.authService.getAuthenticatedUserFromStore().pipe(
+            switchMap((user) => {
+              const updatedUser = cloneDeep(user);
+              Metadata.setFirstValue(updatedUser.metadata, END_USER_AGREEMENT_METADATA_FIELD, String(accepted));
+              return this.ePersonService.update(updatedUser);
```

Comment:
Hello! I have verified that using this dataService's update method it does not wait for the backend response to the PATCH request but the ePerson present in the cache is returned directly. I'm wrong? In this way, therefore, the response coming back is alwyas success even if there may have been problems on the backend side.

Reply:
I think you're right. We'll rewrite it using the patch method instead.